addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryHelper.java

```diff
@@ -0,0 +1,228 @@
+ *
+ * The helper class will open the given serial port and wait for telegrams.
+ * After SERIAL_PORT_AUTO_DETECT_TIMEOUT seconds it will switch the baud rate and wait again for telegrams.
+ * After DSMR_DISCOVERY_TIMEOUT seconds the helper will give up (assuming no DSMR Bridge is present)
+ *
+ * If a telegram is received with at least 1 Cosem Object a bridge is assumed available and a Thing is added
+ * (regardless if there were problems receiving the telegram) and the discovery is stopped.
+ *
+ * If there are commmunication problems the baud rate is switched
+ *
+ * @author M. Volaart
+ * @since 2.1.0
+ */
+public class DSMRBridgeDiscoveryHelper implements DSMRPortEventListener {
+    // logger
+    private final Logger logger = LoggerFactory.getLogger(DSMRBridgeDiscoveryHelper.class);
+
+    // List of available DiscoveryEvents
+    private enum DiscoveryEvent {
+        DISCOVERY_START,
+        DISCOVERY_FAILED,
+        DISCOVERY_SWITCHBAUDRATE,
+        BRIDGE_DISCOVERED;
+    }
+
+    // Discovery state
+    private enum DiscoveryState {
+        DISCOVERY_RUNNING,
+        DISCOVERY_FINISHED
+    }
+
+    // Discovery status (also used for locking the state object)
+    private class DiscoveryStatus {
+        DiscoveryState state;
+    }
+
+    // The port name
+    private final String portName;
+
+    // DSMR Port instance
+    private DSMRPort dsmrPort;
+
+    // current status
+    private final DiscoveryStatus status;
+
+    // Listener for discovered devices
+    private final DSMRBridgeDiscoveryListener discoveryListener;
+
+    // Service for handling timers
+    private ScheduledExecutorService discoveryTimers = ThreadPoolManager
+            .getScheduledPool(DSMRBindingConstants.DSMR_SCHEDULED_THREAD_POOL_NAME);
+
+    /**
+     * Creates a new DSMRBridgeDiscoveryHelper
+     *
+     * @param portName the port name (e.g. /dev/ttyUSB0 or COM1)
+     * @param listener the {@link DSMRMeterDiscoveryListener} to notify of new detected bridges
+     */
+    public DSMRBridgeDiscoveryHelper(String portName, DSMRBridgeDiscoveryListener listener) {
+        status = new DiscoveryStatus();
+        status.state = DiscoveryState.DISCOVERY_RUNNING;
+        this.portName = portName;
+        this.discoveryListener = listener;
+    }
+
+    /**
+     * Start the discovery
+     */
+    public void startDiscovery() {
+        logger.debug("Start discovery for port {}", portName);
+        handleDiscoveryEvent(DiscoveryEvent.DISCOVERY_START);
+    }
+
+    /**
+     * Handle discovery events
+     *
+     * @param event {@link DiscoveryEvent}
+     */
+    private void handleDiscoveryEvent(DiscoveryEvent event) {
+        logger.debug("Handle discovery event {} in state {}", event, status.state);
+        synchronized (status) {
+            if (status.state == DiscoveryState.DISCOVERY_RUNNING) {
+                switch (event) {
+                    case DISCOVERY_START:
+                        discoveryTimers.schedule(new Runnable() {
```

Comment:
Can these be scheduled in the normal scheduler or do you need a lot of them or let them run continuously?